### ISSUE TITLE
Discussion: Automatic transactions for insertGraph

### ISSUE THREAD

**Issue by joepie91** [Vincit/objection.js#1939]
In the `insertGraph` documentation, the following is stated:

> insertGraph operation is not atomic by default! You need to start a transaction and pass it to the query using any of the supported ways. See the section about transactions for more information.

However, I was unable to find any further information about the rationale behind this design decision; the only thing I found was an issue which stated "Objection will never automatically start a transaction", but without explaining why that is.

I feel that this is an ergonomic oversight, personally. It's difficult to imagine a case where one wants to do a graph insert, which conceptually is a single operation, *without* wanting atomicity for the inserted records in question. Which means that, in practice, users will end up doing one of two things the majority of the time:

1. Forget to add a transaction, and have non-atomic graph inserts when they wanted *atomic* inserts. This is a bug that could lead to an inconsistent state of the database.
2. Wrap every graph insert in a transaction, either ad-hoc in each place, or using a custom wrapper function that isn't a part of Objection's API. This is awkward and often poorly maintainable.

At the same time, I don't see any obvious reasons why Objection *shouldn't* start a transaction here; while it's not a widely-known feature, "savepoints" are a part of the SQL standard and should therefore be supported by all SQL databases. They allow for constructing conceptual "subtransactions" within an existing transaction, meaning that automatically creating a transaction for a graph insert would not interfere with existing transaction code.

Knex has full support for these, automatically creating and committing savepoints when calling `.transaction` on an existing transaction object, and so "subtransactions" are possible here without additional effort, and without interfering with application code, as far as I can tell.

I would therefore like to propose enabling automatic transaction creation for graph inserts (possibly with an opt-out), or at least providing a configuration method to opt into this behaviour, as it would save a lot of boilerplate. The former would be preferable due to being safe-by-default, but the latter would at least solve the ergonomics problem.

If there's some reason why this isn't viable, I'd also be interested in further discussion about that!

**Comment by koskimas** [comment]
Objection doesn't automatically start transactions because the user wouldn't have any control over them. What if the user wants to change the isolation level of that transaction? What if the user wants to run multiple operations inside the same transaction? Savepoints could help, but is that always what the user wants? If the user starts a transaction and runs multiple operations in it in addition to an `insertGraph` operation, does the user always want to only roll back the `insertGraph` if it fails, and not the whole transaction? I don't think so.

If we were to start automatic transactions under the hood, we would need to provide APIs for controlling/disabling them, and that would be yet another thing to develop and maintain. It's just easier and cleaner to not start implicit transactions.

**Comment by joepie91** [comment]
> What if the user wants to change the isolation level of that transaction? 

That's a reasonable concern. However, this in and of itself would be possible to make a part of the API, I think - either with 1) an explicit per-query configuration option for the isolation level, 2) a "run in transaction before insert" hook, or 3) simply an instruction that the user should start their own wrapper transaction first, if and when they want this.

Since savepoints do not change the isolation level in and of themselves, the automatically-created transaction in the `insertGraph` operation should not interfere with the third approach at all.

I would personally probably tend towards option 3 + a 'global' setting or hook of some sort for the default isolation level (if one does not exist yet), unless it turns out to be particularly easy to implement option 1. This would keep API verbosity and surface to a minimum, while still only making manual transaction wrapping necessary when the developer needs to deviate from whatever their default isolation level is.

> What if the user wants to run multiple operations inside the same transaction? Savepoints could help, but is that always what the user wants? If the user starts a transaction and runs multiple operations in it in addition to an insertGraph operation, does the user always want to only roll back the insertGraph if it fails, and not the whole transaction? I don't think so.

The way that transaction semantics in Knex work by default, is that if a "subtransaction" fails, it is rolled back to its corresponding savepoint; and the error is then propagated outwards for the calling code to handle. If this handling does not occur, it will propagate up to the parent transaction and roll back *that* one too, and so on, and so forth, all the way until the top-level transaction.

This means that an implicit transaction in `insertGraph` would still end up leaving it to the user how to handle a failure there, same as *without* the implicit transaction. If they catch and handle the error right above the `insertGraph` call, then a partial rollback occurs. If they do not, then a full rollback occurs as the error continues travelling up the transaction tree.

Considering those semantics, I can't really think of a multiple-operation case where savepoints get in the way; they are purely additive, and don't change or interfere with any of the existing behaviour of the calling code. Their effect is contained to the part of the transaction that is contained within it, which would be fully internal to `insertGraph`.

The only edgecase that I can see might be "I explicitly don't want this graph insert to occur within a transaction, I want failures to be tolerated" - which, if people have a concrete reason to need that, can be accomplished with an opt-out from the implicit transaction. I doubt that this will be at all common, though. I have difficulty coming up with even a single realistic case for this.

> If we were to start automatic transactions under the hood, we would need to provide APIs for controlling/disabling them, and that would be yet another thing to develop and maintain. It's just easier and cleaner to not start implicit transactions.

I don't agree that it is "cleaner". It reduces the complexity within Objection itself, yes; but it does so by moving that complexity outwards to the application, where it is much more difficult to centralize that complexity in an ergonomic way than it would have been within Objection. The "waterbed theory" comes to mind here.

Considering how widespread this requirement is, even if a lot of developers probably don't realize that they have it, I don't think that's a good tradeoff, personally.

**Comment by fiznool** [comment]
Just a note to mention that I was surprised when `insertGraph` and friends isn't atomic by default. I totally understand the concerns of starting transactions automatically, however could the following approach alleviate these issues?

If a transaction is passed to the query builder, use it for `insertGraph` and friends. Otherwise, use an implicit one?

So for example, this would use the explicitly created transaction:

``` ts
await Model.transaction(async (trx) => {
  const jennifer = await Person.query(trx).insertGraphAndFetch({
      firstName: 'Jennifer',
      lastName: 'Lawrence',
      pets: [{ 
        name: 'Scrappy' 
      }]
    });
  return jennifer;
});
```

However, this would implicitly create a transaction, just for the `insertGraphAndFetch` operation:

``` ts
const jennifer = await Person.query().insertGraphAndFetch({
  firstName: 'Jennifer',
  lastName: 'Lawrence',
  pets: [{ 
    name: 'Scrappy' 
  }]
});
```

**Comment by lehni** [comment]
I think the current behavior is by design. I am closing this as a wontfix.

**Comment by BlairCurrey** [comment]
I just found this post when searching for whether or not `insertGraphAndFetch` uses a transaction internally. I guessed that it would, but I see from this thread that it doesn't. I was curious exactly how partial failures looked in practice so I created this little demo: https://github.com/BlairCurrey/objection-insertgraph-trx 

In hindsight I probably should have expected the results from the docs, but it was still ambiguous to me until I observed it in action.

The gist of it is a `User` model with many `Posts` will insert everything but the bad post from this example, and an error will be thrown:

```ts
const newUser = await User.query().insertGraphAndFetch({
  name: "Alice",
  posts: [
    {
      title: "Title1",
      body: "This is the body of the first post",
    },
    // This title is not unique and will fail the db's unique constraint
    {
      title: "Title1",
      body: "This is the body of the second post",
    },
    {
      title: "Title3",
      body: "This is the body of the third post",
    },
  ],
}); // this errors
```

I can see both perspectives on this one but tend to think making it atomic is the more normal use-case so ideally the interface would reflect that. But I share more for informational purposes.
